### ISSUE TITLE
[ENHANCEMENT] Harvester Infrastructure+Cloud Provider IaC provisioning

### ISSUE THREAD

**Issue by ionfury** [harvester/harvester#5492]
**Is your enhancement related to a problem? Please describe.**
Deploying RKE2 clusters with Terraform is my primary use case for Harvester.  [Here](https://registry.terraform.io/providers/rancher/rancher2/latest/docs/resources/cluster_v2#create-a-node-driver-cluster-with-harvester-as-both-the-infrastructure-provider-and-cloud-provider) are the provider documents for doing so with Rancher.  Having a curl callout in this process is an anti-pattern in Terraform.  This is a pain to work around.

**Describe the solution you'd like**
The ability to provision and use service accounts for the Harvester Cloud Provider entirely Terraform.

**Describe alternatives you've considered**
[Workaround](https://github.com/ionfury/homelab/blob/1318483042a07225348af50ba81b14329b3ecb1e/terraform/.modules/rancher-harvester-cluster/main.tf#L27).

**Additional context**
The Harvester documentation presents GUI management as the primary use/case selling point.  I'm **much** more interested in the GitOps/IaC management opportunities in this product which feel like a second class citizen.  Given how k8s native Rancher/Harvester are, this strikes me as a missed opportunity.

**Comment by bcdurden** [comment]
I have also encountered issues with some of the non-K8S API within Harvester. However, I think the method of creating downstream RKE2 clusters within Harvester is best done using Rancher's node-provisioner (and thus the cattle-agent), not Terraform. MCM exposes a good deal of features that affect the cluster as a whole (as well as being able to constantly monitor/manage the cluster and node state) as opposed to building individual nodes using Terraform.

See the helmcharts here: https://github.com/rancherfederal/rancher-cluster-templates  as they will allow you to build harvester-hosted RKE2/K3S clusters using the Rancher CRD backend that leverages the node-provisioner.

However, either method still has the same core problem of needing the scoped kubeconfig to be generated if you want to use the cloud-provider. And this requires a RESTful request made to the `https://${RANCHER_URL}/v3/clusters/${CLUSTER_ID}?action=generateKubeconfig` endpoint. Which also requires having to discover the harvester cluster id (a generative name)

If one uses the UI instead to create the cloud-provider credential, it ALSO uses a generative name which becomes impossible to leverage in GitOps as it introduces a chicken/egg situation (you don't know the credential name until after it is created and then have to modify the cluster spec afterwards). The only way I got around it was by creating the credential manually via `curl` and feeding that into a k8s secret with appropriate labels so it did not have a generated name.

Regarding the additional context:
```The Harvester documentation presents GUI management as the primary use/case selling point. I'm much more interested in the GitOps/IaC management opportunities in this product which feel like a second class citizen. Given how k8s native Rancher/Harvester are, this strikes me as a missed opportunity.```

I agree with this. However, I think the primary entry point for GitOps and IaC is done via Rancher MCM managing Harvester itself as the higher-level CRDs describing clusters as objects along with CAPI support is introduced. The new rancher on vcluster option makes this a pretty lightweight option too.

**Comment by ionfury** [comment]
As far as I'm aware the Rancher terraform provider is just a layer on top of those higher-level Rancher Cluster CRDs.  At the time I was implementing Rancher+Harvester I was not able to find any documentation on using and applying the CRDs directly inside Rancher and opted to use the terraform provider instead.  If managing the CRDs directly on the Rancher cluster is the recommended approach to GitOps I would gladly adopt it.

I have also encountered the same difficulties handling the generated cluster id through GitOps.  I think this is a less of an issue using the terraform approach, as you can reference values which are generated at runtime, but the necessity to capture and save this generated value rather than being able to use a given name for the cluster is a pain.

**Comment by bcdurden** [comment]
> As far as I'm aware the Rancher terraform provider is just a layer on top of those higher-level Rancher Cluster CRDs.

The Rancher terraform provider is, yes (rancher2). The Harvester terraform provider is a different entity and is a bit more low-level.  The flow I typically suggest is:
1) Bootstrap and Configure Harvester (networks, VM images, etc)
2) Using Terraform, provision a stand-alone RKE2 cluster onto Harvester.
3) Install Rancher onto RKE2 cluster using Helm via Terraform or bash.
4) Bootstrap and configure Rancher, add Harvester cluster to Rancher's virt management, create cloud creds, etc
5) (a) Create downstream RKE2/K3S clusters using Helm via bash or via Fleet/Flux/Argo
     (b) Create apps in Fleet (or chosen gitops/cd orchestrator like Flux, Argo, etc) with matching labels to downstream clusters

Given that you can manage both the applications going into the downstream clusters as well as the clusters themselves from the same GitOps processes (using built-in Fleet no less for simpler use-cases), it becomes a pretty compelling argument (imo) for using the CRDs directly. Unified APIs for all things in the environment around K8S.

**Comment by ionfury** [comment]
That's the same pattern I'm following with the exception of using the [rancher2](https://registry.terraform.io/providers/rancher/rancher2/latest/docs) terraform provider for cluster provisioning, not the harvester terraform provider.  Perhaps this issue would be more appropriate in the Rancher repo to avoid confusion.

**Comment by bcdurden** [comment]
Yeah I would raise an issue there. There's magic stuff that happens behind the Rancher MCM API that manages what you're asking.
